tensorboard/webapp/plugins/npmi/views/violin_filters/violin_filter/violin_filter_component.ts

```diff
@@ -0,0 +1,376 @@
+  private xAxis?: d3.Axis<string>;
+  private yScale: d3.ScaleLinear<number, number> = d3.scaleLinear();
+  private yAxis?: d3.Axis<number | {valueOf(): number}>;
+  private xScaleNum: d3.ScaleLinear<number, number> = d3.scaleLinear();
+  // Brush
+  private brush: d3.BrushBehavior<unknown> = d3.brushY();
+  // Misc
+  private nanLine!: d3.Selection<SVGLineElement, unknown, null, undefined>;
+  private nanText!: d3.Selection<SVGTextElement, unknown, null, undefined>;
+  private zeroLine!: d3.Selection<SVGLineElement, unknown, null, undefined>;
+
+  private maxBinSize = 0;
+  private readonly rgbColors = ['240, 120, 80', '46, 119, 182', '190, 64, 36'];
+  private readonly area = d3
+    .area<ViolinBin>()
+    .x0(
+      function (this: ViolinFilterComponent, d: ViolinBin) {
+        return this.xScaleNum(-d.length);
+      }.bind(this)
+    )
+    .x1(
+      function (this: ViolinFilterComponent, d: ViolinBin) {
+        return this.xScaleNum(d.length);
+      }.bind(this)
+    )
+    .y(
+      function (this: ViolinFilterComponent, d: ViolinBin) {
+        if (d.x0! === -Infinity) {
+          return this.chartHeight - this.drawMargin.top;
+        }
+        return this.yScale((d.x1! + d.x0!) / 2.0);
+      }.bind(this)
+    )
+    .curve(d3.curveCatmullRom);
+
+  ngAfterViewInit(): void {
+    this.svg = d3.select(this.chartContainer.nativeElement).select('svg');
+    this.svg;
+    this.mainContainer = this.svg
+      .append('g')
+      .attr('transform', `translate(${this.margin.left}, ${this.margin.top})`);
+    this.drawContainer = this.mainContainer
+      .append('g')
+      .attr(
+        'transform',
+        `translate(${this.drawMargin.left}, ${this.drawMargin.top})`
+      );
+    this.dotsGroup = this.drawContainer.append('g').attr('class', 'dotsGroup');
+    this.yAxisGroup = this.mainContainer
+      .append('g')
+      .attr('class', 'axis axis--y');
+    this.xAxisGroup = this.mainContainer
+      .append('g')
+      .attr('class', 'axis axis--x');
+    this.miscGroup = this.drawContainer.append('g');
+    this.xScale = d3.scaleBand().padding(0.05);
+    this.yScale = d3.scaleLinear().range([this.drawHeight, 0]);
+    this.xScaleNum = d3.scaleLinear();
+    this.initializeBrush();
+    this.drawMisc();
+    this.redraw();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (this.svg) {
+      this.redraw();
+    }
+  }
+
+  redraw() {
+    this.setMaxBinSize();
+    this.updateAxes();
+    this.draw();
+  }
+
+  // Initializing/Updating the visualization props
+  setMaxBinSize() {
+    Object.values(this.chartData.violinData).forEach((dataElement) => {
+      const lengths = dataElement.map((bin) => bin.length);
+      const longest: number = Math.max(...lengths);
+      this.maxBinSize = Math.max(longest, this.maxBinSize);
+    });
+  }
+
+  updateAxes() {
+    this.xScale
+      .range([0, this.drawWidth])
+      .domain(Object.keys(this.chartData.violinData));
+    this.xAxis = d3.axisBottom(this.xScale);
```

Comment:
I guess I was not being explicit: do you also need to set the scale on axis over and over or are they capable of detecting changes and make DOM changes accordingly?

Reply:
Ahhh. Apparently, yes. Changed that, thanks.